File: packages/code-review/tools/review-ingest-pr-decisions.py
#!/usr/bin/env python3
"""Ingere decisões de review a partir de um PR mergeado (comentários /avaliar-inline).

Heurísticas (ver lib/ingest_decisions.py):
- Resposta humana rejeitando o achado → rejeitado / nao-aplicavel → exclusions.yaml
- Resposta humana sem objeção → aceito → candidate → convencoes.md (≥2×)
- Fix aplicado no merge (suggestion / De / intra-PR) sem reply → aceito
- Merge sem reply e achado ainda presente → rejeitado → exclusions.yaml

CLI + I/O (gh/git). Regras puras em lib.ingest_decisions.
"""
from __future__ import annotations

import argparse
import importlib.util
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

_TOOLS = Path(__file__).resolve().parent
if str(_TOOLS) not in sys.path:
    sys.path.insert(0, str(_TOOLS))

from lib.ingest_decisions import (  # noqa: E402
    classify_thread,
    extract_de_para_from_body,
    extract_summary,
    parse_repo,
    resolve_pr_commit_range,
    upsert_pr_decisions,
)


def _load_memoria():
    spec = importlib.util.spec_from_file_location(
        "review_memoria", _TOOLS / "review-memoria.py"
    )
    if spec is None or spec.loader is None:
        raise RuntimeError("review-memoria.py não encontrado")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


_mem = _load_memoria()
DECISIONS_INGEST_FILE = _mem.DECISIONS_INGEST_FILE
build_context = _mem.build_context
cmd_promover = _mem.cmd_promover
decisions_ingest_path = _mem.decisions_ingest_path
read_decisions = _mem.read_decisions
review_dir = _mem.review_dir
write_context = _mem.write_context


def run_gh(args: list[str]) -> Any:
    proc = subprocess.run(
        ["gh", *args],
        check=True,
        capture_output=True,
        text=True,
    )
    return json.loads(proc.stdout or "null")


def fetch_pr_threads(owner: str, repo: str, pr_number: int) -> dict[str, Any]:
    query = """
query($owner: String!, $repo: String!, $number: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $number) {
      merged
      mergedAt
      baseRefOid
      headRefOid
      mergeCommit { oid }
      title
      reviewThreads(first: 100) {
        nodes {
          isResolved
          comments(first: 30) {
            nodes {
              body
              path
              line
              originalLine
              author { login }
              commit { oid }
            }
          }
        }
      }
    }
  }
}
"""
    data = run_gh(
        [
            "api",
            "graphql",
            "-f",
            f"query={query}",
            "-f",
            f"owner={owner}",
            "-f",
            f"repo={repo}",
            "-F",
            f"number={pr_number}",
        ]
    )
    pr = data.get("data", {}).get("repository", {}).get("pullRequest")
    if not pr:
        raise SystemExit(f"PR #{pr_number} não encontrado em {owner}/{repo}")
    return pr


def run_export_exclusions(project: Path) -> None:
    script = _TOOLS / "review-export-exclusions.sh"
    subprocess.run(["bash", str(script), str(project)], check=True)


def cmd_ingest(
    project: Path,
    pr_number: int,
    repository: str,
    write: bool,
    promote_all: bool,
) -> int:
    try:
        owner, repo = parse_repo(repository)
    except ValueError as exc:
        raise SystemExit(str(exc)) from exc

    pr = fetch_pr_threads(owner, repo, pr_number)

    if not pr.get("merged"):
        print(f"PR #{pr_number} não foi mergeado — nada a ingerir.")
        return 0

    merge_oid = (pr.get("mergeCommit") or {}).get("oid") or ""
    base_oid, head_oid = resolve_pr_commit_range(project, pr, merge_oid)
    threads = (pr.get("reviewThreads") or {}).get("nodes") or []

    proposed: list[dict[str, Any]] = []
    for thread in threads:
        decision = classify_thread(
            thread, True, merge_oid, base_oid, head_oid, project, pr_number
        )
        if decision:
            proposed.append(decision)

    print(f"=== ingest PR #{pr_number} ({owner}/{repo}) ===")
    print(f"Merge: {merge_oid[:7] if merge_oid else '?'}")
    print(f"Threads /avaliar: {len(proposed)} decisão(ões) proposta(s)")
    print("")

    for d in proposed:
        print(
            f"  [{d['decision']}] {d.get('file') or '?'}:{d.get('line')} — "
            f"{d.get('summary', '')[:70]}"
        )

    if not proposed:
        print("\nNenhuma decisão nova.")
        return 0

    if not write:
        print("\nDry-run. Use --write para gravar decisions + compactar + promover + export.")
        return 0

    rd = review_dir(project)
    rd.mkdir(parents=True, exist_ok=True)
    version = rd / ".memoria-version"
    if not version.is_file():
        version.write_text("2\n", encoding="utf-8")

    decisions_path = decisions_ingest_path(project)
    added = upsert_pr_decisions(decisions_path, pr_number, proposed)
    print(
        f"\n{len(added)} decisão(ões) gravada(s) em {DECISIONS_INGEST_FILE} "
        f"(source github-pr-{pr_number})"
    )

    all_decisions = read_decisions(decisions_path)
    context = build_context(project, f"github-pr-{pr_number}", all_decisions)
    write_context(rd, context)
    print("context.yaml atualizado")

    cmd_promover(project, write=True, all_candidates=promote_all)

    run_export_exclusions(project)
    print("exclusions.yaml exportado")
    print(
        "\nPróximo: commit .cursor/review/decisions-ingest.jsonl + "
        "convencoes.md + exclusions.yaml"
    )
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description="Ingere decisões de PR mergeado → memória v2")
    parser.add_argument("pr_number", type=int, help="Número do PR mergeado")
    parser.add_argument("--project", default=".", help="Raiz do repo alvo")
    parser.add_argument("--repo", default="", help="owner/name (default: gh repo view)")
    parser.add_argument("--write", action="store_true", help="Persistir memória")
    parser.add_argument(
        "--all",
        action="store_true",
        help="Promover candidates com 1 ocorrência (default: ≥2)",
    )
    args = parser.parse_args()

    project = Path(args.project).resolve()
    repository = args.repo.strip()
    if not repository:
        proc = subprocess.run(
            ["gh", "repo", "view", "--json", "nameWithOwner", "-q", ".nameWithOwner"],
            cwd=project,
            capture_output=True,
            text=True,
            check=True,
        )
        repository = proc.stdout.strip()

    return cmd_ingest(project, args.pr_number, repository, args.write, args.all)


if __name__ == "__main__":
    sys.exit(main())

File: packages/code-review/bin/review-ingest-pr-decisions.ts
#!/usr/bin/env node
/**
 * Ingere decisões de review a partir de um PR mergeado.
 */
import { spawnSync } from "node:child_process";
import { mkdirSync, writeFileSync, existsSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import {
  classifyThread,
  parseRepo,
  resolvePrCommitRange,
  upsertPrDecisions,
} from "../src/ingest/index.js";
import {
  DECISIONS_INGEST_FILE,
  buildContext,
  cmdPromover,
  decisionsIngestPath,
  readDecisions,
  reviewDir,
  writeContext,
} from "../src/memory/index.js";
import {
  dualWriteDecisions,
  logDualWriteResult,
} from "../src/store/index.js";
import { resolveProjectSlug } from "../src/store/project-slug.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const toolsDir = path.resolve(__dirname, "../tools");

function runGh(args: string[]): unknown {
  const proc = spawnSync("gh", args, {
    encoding: "utf8",
  });
  if (proc.status !== 0) {
    throw new Error(proc.stderr || `gh failed: ${args.join(" ")}`);
  }
  return JSON.parse(proc.stdout || "null");
}

function fetchPrThreads(
  owner: string,
  repo: string,
  prNumber: number
): Record<string, unknown> {
  const query = `
query($owner: String!, $repo: String!, $number: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $number) {
      merged
      mergedAt
      baseRefOid
      headRefOid
      mergeCommit { oid }
      title
      reviewThreads(first: 100) {
        nodes {
          isResolved
          comments(first: 30) {
            nodes {
              body
              path
              line
              originalLine
              author { login }
              commit { oid }
            }
          }
        }
      }
    }
  }
}
`;
  const data = runGh([
    "api",
    "graphql",
    "-f",
    `query=${query}`,
    "-f",
    `owner=${owner}`,
    "-f",
    `repo=${repo}`,
    "-F",
    `number=${prNumber}`,
  ]) as {
    data?: { repository?: { pullRequest?: Record<string, unknown> } };
  };
  const pr = data?.data?.repository?.pullRequest;
  if (!pr) {
    console.error(`PR #${prNumber} não encontrado em ${owner}/${repo}`);
    process.exit(1);
  }
  return pr;
}

function runExportExclusions(project: string): void {
  const script = path.join(toolsDir, "review-export-exclusions.sh");
  const proc = spawnSync("bash", [script, project], { encoding: "utf8" });
  if (proc.status !== 0) {
    throw new Error(proc.stderr || "review-export-exclusions.sh falhou");
  }
}

async function cmdIngest(
  project: string,
  prNumber: number,
  repository: string,
  write: boolean,
  promoteAll: boolean
): Promise<number> {
  let owner: string;
  let repo: string;
  try {
    [owner, repo] = parseRepo(repository);
  } catch (err) {
    console.error(err instanceof Error ? err.message : String(err));
    process.exit(1);
  }

  const pr = fetchPrThreads(owner, repo, prNumber);

  if (!pr.merged) {
    console.log(`PR #${prNumber} não foi mergeado — nada a ingerir.`);
    return 0;
  }

  const mergeOid =
    ((pr.mergeCommit as { oid?: string } | null)?.oid as string) || "";
  const [baseOid, headOid] = resolvePrCommitRange(project, pr, mergeOid);
  const threads =
    (
      (pr.reviewThreads as { nodes?: unknown[] } | null)?.nodes as Array<{
        isResolved?: boolean;
        comments?: { nodes?: unknown[] };
      }>
    ) ?? [];

  const proposed: Array<Record<string, unknown>> = [];
  for (const thread of threads) {
    const decision = classifyThread(
      thread as Parameters<typeof classifyThread>[0],
      true,
      mergeOid,
      baseOid,
      headOid,
      project,
      prNumber
    );
    if (decision) proposed.push(decision);
  }

  console.log(`=== ingest PR #${prNumber} (${owner}/${repo}) ===`);
  console.log(`Merge: ${mergeOid ? mergeOid.slice(0, 7) : "?"}`);
  console.log(`Threads /avaliar: ${proposed.length} decisão(ões) proposta(s)`);
  console.log("");

  for (const d of proposed) {
    console.log(
      `  [${d.decision}] ${d.file || "?"}:${d.line} — ${String(d.summary ?? "").slice(0, 70)}`
    );
  }

  if (!proposed.length) {
    console.log("\nNenhuma decisão nova.");
    return 0;
  }

  if (!write) {
    console.log(
      "\nDry-run. Use --write para gravar decisions + compactar + promover + export."
    );
    return 0;
  }

  const rd = reviewDir(project);
  mkdirSync(rd, { recursive: true });
  const version = path.join(rd, ".memoria-version");
  if (!existsSync(version)) writeFileSync(version, "2\n", "utf8");

  const decisionsPath = decisionsIngestPath(project);
  const added = upsertPrDecisions(decisionsPath, prNumber, proposed);
  console.log(
    `\n${added.length} decisão(ões) gravada(s) em ${DECISIONS_INGEST_FILE} (source github-pr-${prNumber})`
  );

  const dual = await dualWriteDecisions(added, {
    projectSlug:
      resolveProjectSlug(project) ||
      String(process.env.REVIEW_PROJECT_SLUG ?? "").trim() ||
      undefined,
    run: {
      source: "ci",
      prNumber,
      reviewSlug: `pr-${prNumber}`,
      actorKind: "tool",
      actorRef: "review-ingest-pr",
      meta: { kind: "pr-ingest" },
    },
    decidedBy: "review-ingest-pr",
  });
  logDualWriteResult("review-ingest-pr", dual);
  if (dual.error) {
    console.error("Ingest: falha ao gravar no store — abortando");
    return 1;
  }

  const allDecisions = readDecisions(decisionsPath);
  const context = buildContext(
    project,
    `github-pr-${prNumber}`,
    allDecisions
  );
  writeContext(rd, context);
  console.log("context.yaml atualizado (cache local)");

  cmdPromover(project, true, promoteAll);

  runExportExclusions(project);
  console.log("exclusions.yaml exportado (cache local; fonte de verdade = store)");
  return 0;
}

async function main(): Promise<number> {
  const argv = process.argv.slice(2);
  let project = ".";
  let write = false;
  let promoteAll = false;
  let repo = "";
  let prNumber: number | null = null;

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--project" && argv[i + 1]) {
      project = argv[++i];
    } else if (a === "--repo" && argv[i + 1]) {
      repo = argv[++i];
    } else if (a === "--write") {
      write = true;
    } else if (a === "--all") {
      promoteAll = true;
    } else if (/^\d+$/.test(a)) {
      prNumber = Number.parseInt(a, 10);
    }
  }

  if (prNumber === null) {
    console.error(
      "Uso: review-ingest-pr-decisions <pr_number> [--project DIR] [--repo owner/name] [--write] [--all]"
    );
    return 1;
  }

  const resolved = path.resolve(project);
  let repository = repo.trim();
  if (!repository) {
    const proc = spawnSync(
      "gh",
      ["repo", "view", "--json", "nameWithOwner", "-q", ".nameWithOwner"],
      { cwd: resolved, encoding: "utf8" }
    );
    if (proc.status !== 0) {
      console.error(proc.stderr || "gh repo view falhou");
      return 1;
    }
    repository = (proc.stdout ?? "").trim();
  }

  return cmdIngest(resolved, prNumber, repository, write, promoteAll);
}

main().then(
  (code) => process.exit(code),
  (err) => {
    console.error(err instanceof Error ? err.message : err);
    process.exit(1);
  }
);
